gbrl/common/utils.py

```
@@ -594,3 +605,108 @@ def ensure_leaf_tensor_or_array(array: NumericalData,
+    direction_map = {
+        "increasing": 1, "inc": 1, "+": 1, 1: 1,
+        "decreasing": -1, "dec": -1, "-": -1, -1: -1
+    }
+
+    for feat_idx, (direction, output_dims) in constraints.items():
+        # Validate feature index
+        if not isinstance(feat_idx, int) or feat_idx < 0 or feat_idx >= input_dim:
+            raise ValueError(
+                f"Invalid feature index {feat_idx}. "
+                f"Must be an integer in [0, {input_dim})"
+            )
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)